### PR TITLE
[catmem] Use concurrent ring buffer

### PR DESCRIPTION
**Pull request by iyzhang**
This PR moves Catmem to use the concurrent ring buffer.

### CODE REVIEW

File: src/rust/collections/concurrent_ring.rs [hunk 1]
```diff
@@ -307,7 +307,7 @@ impl ConcurrentRingBuffer {
         #[cfg(feature = "profiler")]
         timer!("collections::concurrent_ring::release_space");
         let len_: usize = align_header(len + HEADER_SIZE);
-        let new_offset: usize = current_offset + len_ % self.capacity();
+        let new_offset: usize = (current_offset + len_) % self.capacity();
```
Comment (ppenna) [inline src/rust/collections/concurrent_ring.rs:310]: Good catch!

File: src/rust/catmem/ring.rs [hunk 2]
```diff
@@ -80,48 +81,40 @@ impl Ring {
...
-            mutex: Mutex::new(),
         })
     }
 
     /// Try to pop a byte from the shared memory ring. If successful, return the byte and whether the eof flag is set,
     /// otherwise return None for a retry.
-    pub fn try_pop(&mut self) -> Result<(Option<u8>, bool), Fail> {
-        self.state_machine.may_pop()?;
+    pub fn try_pop(&mut self, buf: &mut [u8]) -> Result<(usize, bool), Fail> {
+        let mut msg: Vec<u8> = vec![0; buf.len() + mem::size_of::<u32>()];
```
Comment (ppenna) [inline src/rust/catmem/ring.rs:90]: Instead of `mem::size_of::<u32>()`, shouldn't it be `HEADER_SIZE`?

File: src/rust/catmem/ring.rs [hunk 3]
```diff
@@ -142,18 +135,11 @@ impl Ring {
...
-        }
-
-        let result: Result<(), u16> = self.push_buf.try_enqueue(EOF);
-        assert_eq!(self.mutex.unlock().is_ok(), true);
-
-        match result {
-            Ok(()) => Ok(()),
+        match self.push_buf.try_push(&EOF) {
+            Ok(len) => {
+                debug_assert_eq!(len, mem::size_of::<u32>());
```
Comment (ppenna) [inline src/rust/catmem/ring.rs:140]: Instead of `mem:size_of::<u32>`, shouldn't this be `HEADER_SIZE`?

File: src/rust/catmem/queue.rs [hunk 4]
```diff
@@ -340,3 +326,22 @@ impl IoQueue for SharedCatmemQueue {
...
+        return true;
+    }
+
+    #[cfg(target_os = "windows")]
+    if errno == WSAEWOULDBLOCK.0 || errno == WSAEINPROGRESS.0 || errno == WSAEALREADY.0 {
+        return true;
+    }
+
+    return false;
+}
```
Comment (ppenna) [inline src/rust/catmem/queue.rs:347]: This is now provided by the runtime layer.

File: src/rust/catloop/socket.rs [hunk 5]
```diff
@@ -717,10 +717,18 @@ pub fn cook_magic_connect(catmem: &mut SharedCatmemLibOS) -> Result<demi_sgarray
...
-        return Err(e);
+    match len {
+        0 => {
+            let e: Fail = Fail::new(libc::EBADF, "server closed connection");
+            error!("extract_port_number(): failed to establish connection ({:?})", e);
+            return Err(e);
+        },
+        2 => (),
+        _ => {
+            let e: Fail = Fail::new(libc::EINVAL, "server did not send a valid port");
```
Comment (ppenna) [inline src/rust/catloop/socket.rs:728]: `EINVAL` is used for arguments. I would change it for `EBADMSG`

File: src/rust/catmem/ring.rs [hunk 6]
```diff
@@ -80,48 +81,40 @@ impl Ring {
...
             pop_buf: SharedRingBuffer::open(&format!("{}:tx", name), RING_BUFFER_CAPACITY)?,
             state_machine: RingStateMachine::new(),
-            mutex: Mutex::new(),
         })
     }
 
     /// Try to pop a byte from the shared memory ring. If successful, return the byte and whether the eof flag is set,
     /// otherwise return None for a retry.
-    pub fn try_pop(&mut self) -> Result<(Option<u8>, bool), Fail> {
-        self.state_machine.may_pop()?;
```
[the remark this answers is not in the document]
Comment (iyzhang) [reply src/rust/catmem/ring.rs:90]: Looks like you fixed this, so I will resolve

**Review by ppenna — APPROVED** [review]
LGTM

**Pull request merged**